server/app/jobs/grid_service_scheduler_worker.rb

```diff
@@ -9,58 +9,67 @@ def initialize(autostart = true)
 
   def watch
     loop do
-      self.check_deploy_queue
+      if service_deploy = self.check_deploy_queue
+        self.deploy(service_deploy)
+      end
       sleep 1
     end
   end
 
+  # Fetch deploy from queue, and
+  #
+  # @return [GridServiceDeploy, nil] deploy to run
   def check_deploy_queue
     service_deploy = fetch_deploy_item
-    return unless service_deploy
+    return nil unless service_deploy
+
+    if service_deploy.started_at
+      warn "aborting legacy grid_service=#{service_deploy.grid_service_id} deploy from #{service_deploy.created_at}"
+      # legacy deployment before queueing fixes
+      service_deploy.abort! "deploy is legacy"
+      return nil
+    end
+    # TODO: abort stale deploys with ancient created_at, or deploys that have been running for too long
 
     with_dlock("check_deploy_queue:#{service_deploy.grid_service_id}", 10) do
-      service_deploy.reload
-      if service_deploy.grid_service.running? || service_deploy.grid_service.initialized?
-        self.perform(service_deploy)
+      if service_deploy.grid_service.deploy_running?
+        info "delaying #{service_deploy.grid_service.to_path} deploy because there is another deploy in progress"
+        return nil
+
+      elsif service_deploy.grid_service.running? || service_deploy.grid_service.initialized?
+        info "starting #{service_deploy.grid_service.to_path} deploy"
+        service_deploy.set(started_at: Time.now.utc)
+        return service_deploy
+
       else
-        service_deploy.destroy
+        info "aborting #{service_deploy.grid_service.to_path} deploy of non-running service"
+        service_deploy.abort! "service is not running"
+        return nil
       end
     end
   end
 
+  # Mark created/queued deploys as queued, and return for processing.
+  # The caller has 30s to process the returned deploy, or it will be re-fetched.
+  #
   # @return [GridServiceDeploy, NilClass]
   def fetch_deploy_item
-    GridServiceDeploy.where(started_at: nil)
+    GridServiceDeploy.any_of({:_deploy_state => :created}, {:_deploy_state => :queued, :queued_at.lt => 30.seconds.ago})
```

Comment:
This needs both the `deploy_state=queued` as well as the `queued_at` in order to avoid the requirement for an `after(30) { .... }` action, which might never get run. Now the `deploy_state` will normally change to `ongoing` before the `:queued_at.lt => 30.seconds.ago` expires.

Might be able to simplify this if we used a migration to populate the the `deployed_at` for all existing deploys, and this just bumped `queued_at` if the `queued_at` was old and `started_at` was still nil?